src/util/bson.hpp

```diff
@@ -0,0 +1,1218 @@
+                ss << "-Infinity";
+            } else {
+                ss << bson;
+            }
+            ss << '"' << "}";
+        },
+        [&ss](ObjectId bson) {
+            ss << "{" << "\"$oid\"" << ":" << '"' << bson << '"' << "}";
+        },
+        [&ss](BsonDocument bson) {
+            ss << "{";
+            for (auto const& pair : bson)
+            {
+                ss << '"' << pair.first << "\":";
+                to_json(pair.second, ss);
+                ss << ",";
+            }
+            if (bson.size())
+                ss.seekp(-1, std::ios_base::end);
+            ss << "}";
+        },
+        [&ss](BsonArray bson) {
+            ss << "[";
+            for (auto const& b : bson)
+            {
+                to_json(b, ss);
+                ss << ",";
+            }
+            if (bson.size())
+                ss.seekp(-1, std::ios_base::end);
+            ss << "]";
+        },
+        [&ss](BinaryData bson) {
+            ss << "{\"$binary\":{\"base64\":\"" << (bson.is_null() ? "" : bson.data()) << "\",\"subType\":\"00\"}}";
+        },
+        [&ss](RegularExpression bson) {
+            std::stringstream options;
+            for (const auto &option : bson.options()) options << option;
+            ss << "{\"$regularExpression\":{\"pattern\":\"" << bson.pattern()
+                << "\",\"options\":\"" << options.str() << "\"}}";
+        },
+        [&ss](Timestamp bson) {
+            ss << "{\"$timestamp\":{\"t\":" << bson.get_seconds() << ",\"i\":" << 1 << "}}";
+        },
+        [&ss](time_t bson) {
+            ss << "{\"$date\":{\"$numberLong\":\"" << bson << "\"}}";
+        },
+        [&ss](MaxKey) {
+            ss << "{\"$maxKey\":1}";
+        },
+        [&ss](MinKey) {
+            ss << "{\"$minKey\":1}";
+        },
+        [&ss](const char* bson) {
+            ss << '"' << bson << '"';
+        },
+        [&ss](bool bson) {
+            ss << (bson ? "true" : "false");
+        },
+    }, bson);
+}
+} // anonymous namespace
+
+inline std::string to_json(const Bson& bson)
+{
+    std::stringstream ss;
+    to_json(bson, ss);
+    return ss.str();
+}
+
+inline std::string to_json(const BsonDocument& bson)
+{
+    std::stringstream ss;
+    to_json(bson, ss);
+    return ss.str();
+}
+
+
+inline std::ostream& operator<<(std::ostream& out, const Bson& m)
+{
+    out << to_json(m);
+    return out;
+}
+
+/**
+ ======= Equality =======
+ */
+
+template <typename T>
+inline bool operator==(const Bson& lhs, const T& rhs)
+{
+    if (auto value = std::get_if<T>(&lhs))
+        return *value == rhs;
+    return false;
+}
+
+template <>
+inline bool operator==(const Bson& lhs, const std::string& rhs)
+{
+    if (auto value = std::get_if<const char*>(&lhs))
+        return *value == rhs.data();
+    return false;
+}
+
+inline bool operator==(const Bson& lhs, const char rhs[])
+{
+    if (auto value = std::get_if<const char*>(&lhs))
+        return !strcmp(*value, (const char*)rhs);
```

Comment:
this assumes that a null character is the end of the string, so it won't work for strings storing nulls. One fix is to use std::string everywhere, another is to pass in the size of both strings to compare